pennylane/gradients/pulse_gradient.py

```diff
@@ -0,0 +1,529 @@
+            argument doesn't influence the shots used for tape execution, but provides information
+            to the transform about the device shots and helps in determining if a shot sequence
+            was used.
+
+    Returns:
+        function or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, an object representing the Jacobian (function) of the QNode
+          that can be executed to obtain the Jacobian matrix.
+          The type of the matrix returned is either a tensor, a tuple or a
+          nested tuple depending on the nesting structure of the original QNode output.
+
+        - If the input is a tape, a tuple containing a
+          list of generated tapes, together with a post-processing
+          function to be applied to the results of the evaluated tapes
+          in order to obtain the Jacobian matrix.
+
+    This transform realizes the stochastic parameter-shift rule for pulse sequences, as introduced
+    in `Banchi and Crooks (2018) <https://quantum-journal.org/papers/q-2021-01-25-386/>`_ and
+    `Leng et al. (2022) <https://arxiv.org/abs/2210.15812>`_.
+
+    .. note::
+
+        This function requires the JAX interface and does not work with other autodiff interfaces
+        commonly encountered with PennyLane. It additionally only supports the new return type
+        system, see :func:`enable_return`.
+
+    .. note::
+
+        This function uses a basic sampling approach with a uniform distribution to estimate the
+        integral appearing in the stochastic parameter-shift rule. In many cases, there are
+        probability distributions that lead to smaller variances of the estimator.
+        In addition, the sampling approach will not reduce trivially to simpler parameter-shift
+        rules when used with simple pulses (see details and examples below), potentially leading
+        to imprecise results and/or unnecessarily large computational efforts.
+
+    .. note::
+
+        Currently this function only supports pulses for which each *parametrized* term is a
+        simple Pauli word. More general Hamiltonian terms are not supported yet.
+
+    **Examples**
+
+    Consider a pulse program with a single two-qubit pulse, generated by a Hamiltonian
+    with three terms: the non-trainable term :math:`\frac{1}{2}X\otimes I`, the trainable
+    constant (over time) term :math:`v_1 Z\otimes Z` and the trainable sinoidal term
+    :math:`\sin(v_2 t) I\otimes X`.
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        def sin(p, t):
+            return jax.numpy.sin(p * t)
+
+        ZZ = qml.PauliZ(0) @ qml.PauliZ(1)
+        H = 0.5 * qml.PauliX(0) + qml.pulse.constant * ZZ + sin * qml.PauliX(1)
+
+        def ansatz(params):
+            qml.evolve(H)(params, (0.2, 1.))
+            return qml.expval(qml.PauliY(1))
+
+        qnode = qml.QNode(ansatz, dev, interface="jax", diff_method=qml.gradients.stoch_pulse_grad)
```

Comment:
How is it working without QNode integration? Do you plan on adding it later?

Reply:
Yes, I thought this PR is large enough already :sweat_smile: